### DISCUSSION TITLE
Ecosystem Compatibility Tracking (2021-02-10 to 2024-02-05)

### DISCUSSION THREAD

**Discussion by erights** [endojs/endo#1950] [General]
Fear that tape overrode `Function.prototype.apply` cleared by @ljharb at https://github.com/Agoric/SES-shim/issues/474

Good news about JSS and enzyme https://github.com/Agoric/agoric-sdk/pull/2321#issuecomment-777099286

Issue filed with vega https://github.com/vega/vega/issues/3075

**Continues at https://github.com/endojs/endo/issues/2037**

**Comment by erights** [comment]
https://github.com/Agoric/agoric-sdk/pull/2575 patches problems with vega and d3, including the vega problem mentioned above reported at https://github.com/vega/vega/issues/3075

Still need to report problem to d3
Still need to report https://github.com/Agoric/agoric-sdk/issues/2324 to node, that it gets confused if `Object.prototype.constructor` is an accessor property. Attn @bmeck

**Comment by erights** [comment]
https://github.com/vega/vega/pull/3109 would no longer fix https://github.com/vega/vega/issues/3075

**Comment by erights** [comment]
We need to file a bug against https://github.com/feross/buffer that their code is incompat with frozen primordials, causing https://github.com/Agoric/agoric-sdk/issues/2663

**Comment by erights** [comment]
https://github.com/vega/vega/pull/3109 has been merged and https://github.com/vega/vega/issues/3075 closed. Thanks @jheer !

**Comment by erights** [comment]
https://github.com/endojs/endo/issues/621 reports that regenerator-runtime overrides `Object.prototype.constructor` by assignment. We should file a bug against it. In the meantime we will probably patch it.

**Comment by erights** [comment]
More on the regenerator problems
https://github.com/facebook/regenerator/issues/410
https://github.com/facebook/regenerator/issues/411

**Comment by kumavis** [comment]
luxon assigns to `toLocaleString` on an obj :crying_cat_face: https://npmfs.com/package/luxon/1.24.1/build/cjs-browser/luxon.js#L7081
however they are using class syntax in their src https://npmfs.com/package/luxon/1.26.0/src/datetime.js#L1503
so likely a result of using babel https://github.com/moment/luxon/blob/91c5f87e7eed66f09165419455f47809990bd4f8/tasks/build.js#L4
and could be fixed there

**Comment by kumavis** [comment]
`@formatjs/intl-utils` src also has the same `export hasOwnProperty` that ends up as a conflicting `exports.hasOwnProperty` assignment. may be `tsc` performing the esm->cjs transform

**Comment by erights** [comment]
We need to file a bug on React so that https://github.com/endojs/endo/pull/642 is no longer needed to avoid https://github.com/Agoric/dapp-token-economy/issues/159 . If we modify React to fix that bug, we should push that PR upstream.

**Comment by ljharb** [comment]
That last one is private; is there somewhere i can read about the bug?

**Comment by erights** [comment]
https://github.com/endojs/endo/issues/612 is another conflict with react. Attn @rbuckton @ljharb @AidenRourke

**Comment by erights** [comment]
Note that @AidenRourke at #642 says

> If I just have an html file that imports ses and calls lockdown there is no error

We do that as well before any of the rest of react. The need for this is yet another problem with react we need to file.

**Comment by ljharb** [comment]
React uses reflect.metadata?

**Comment by erights** [comment]
> React uses reflect.metadata?

That was within the symptoms we were seeing as well. Specifically, we saw an attempt to add a `Reflect."@decorate"` method. (Not sure I'm getting the name right)

**Comment by rbuckton** [comment]
As far as I can tell, react doesn't use `reflect-metadata` or `Reflect.decorate`. I'd need more to go on to be any help.

**Comment by AidenRourke** [comment]
Hey all, the error I get when trying to use `window.lockdown()` with react is 
```
index.js:1 failed to delete intrinsics.Array.@wry/context:Slot of type function:  (TypeError#1)
```

**Comment by kumavis** [comment]
@AidenRourke is this only react or some additional react packages
[heres what directly uses `@wry/context`](https://www.npmjs.com/browse/depended/@wry/context)

was not familiar with wry :eyes: 
https://github.com/benjamn/wryware
> A collection of packages that are probably a little too clever. Use at your own wrisk.

perhaps too clever, yes

**Comment by rbuckton** [comment]
https://github.com/benjamn/wryware/blob/569d7a6294b77aa9e71d7c7709344aca7cbb75a1/packages/context/src/slot.ts#L126-L135

**Comment by kumavis** [comment]
@AidenRourke this global pollution should gracefully fail if you run lockdown before your dependencies that include `@wry/context`

**Comment by AidenRourke** [comment]
@kumavis btw I think what you're doing with lavamoat is super cool.

To answer your question, I experienced the error even with a stock create react app.

Last time I tried to put the ses import and subsequent call to lockdown in the head of the index.html file, I still got the error.

**Comment by ljharb** [comment]
Any chance you have NODE_OPTIONS set?

**Comment by erights** [comment]
> Any chance you have NODE_OPTIONS set?

We don't. I grepped and the only mention of `NODE_OPTIONS` anywhere in our code is in an `esm-*.patch` file: https://github.com/Agoric/agoric-sdk/blob/master/patches/esm%2B3.2.25.patch

**Comment by ljharb** [comment]
@erights meaning, you can reproduce this issue with a stock CRA, with no NODE_OPTIONS set?

**Comment by michaelfig** [comment]
> @erights meaning, you can reproduce this issue with a stock CRA, with no NODE_OPTIONS set?

That's correct.

**Comment by kumavis** [comment]
>  stock create react app

lots of stuff in a stock CRA
thanks for the report

**Comment by erights** [comment]
See
https://github.com/microsoft/TypeScript/issues/43463
https://github.com/microsoft/TypeScript/issues/43450
https://github.com/mrkmg/node-external-editor/issues/25

**Comment by erights** [comment]
Re https://github.com/endojs/endo/issues/576#issuecomment-809131839 and https://github.com/endojs/endo/issues/612

We need to file a bug against https://github.com/benjamn/wryware/blob/569d7a6294b77aa9e71d7c7709344aca7cbb75a1/packages/context/src/slot.ts#L126-L135 .

**Comment by erights** [comment]
Re https://github.com/vega/vega/pull/3109#issuecomment-792590332
we need to file a bug against rollup for the way it generates the `exports` object by assignment.

**Comment by dckc** [comment]
In https://github.com/Agoric/documentation/pull/579#discussion_r762365040 @erights acknowledges that https://github.com/endojs/endo/wiki also tracks 3rd party package issues and says we should consolidate.

My preference would be to
  - use a separate issue for each compatibility issue
    - it's OK for an issue to start out fuzzy, but in due course it should get a clear scope that can be finished; this issue shows no sign of ever being done.
  - use a github label (perhaps **npm-compat**) to facilitate finding all of them
  - use a wiki page to summarize common problems and solutions with a pointer to the labelled issues
 
Hm... that assumes the resolution to the issues is changes to ses. But I suppose as often as not we're changing the other packages to work with SES. A wiki page makes for a better index than a label, in that case.

OTOH: changes to wiki pages don't result in notifications the way changes to issues do. hm.

**Comment by Jack-Works** [comment]
- immer (writes mapInstance.set) https://github.com/immerjs/immer/pull/914/
- web3-core-method (writes functionInstance.call) https://github.com/ChainSafe/web3.js/pull/4918
- gun (writes so many utility functions) https://github.com/amark/gun/pull/1217
- error-polyfill (try to polyfill Error.captureStackTrace and Error.getStackTrace) https://github.com/inf3rno/error-polyfill/pull/38/

Following packages bundles their own old version of regenerator-runtime (0.11.* or older) which is not compatible with lockdown:
- @ceramicnetwork/rpc-transport
- @ceramicstudio/idx
- @uniswap/v3-sdk
- key-did-provider-ed25519
- rpc-utils (before 0.6.0)

**Comment by Jack-Works** [comment]
maintainer of immer not accepting my PR (link above) 😢

**Comment by dckc** [comment]
The immer maintainer's response suggests that the SES proposal could use some love. Perhaps next time tc39 week rolls around, we could spend some of the less interesting parts of the meeting updating the proposal at least with some comments to let the community know that it is very much active.

> I checked the proposal, but it is still in stage 1 and looks largely inactive for the last 2 years, so I don't see a reason to provide all other users with a slower and slightly less straightforward implementation atm? Typically only at stage 3 it becomes interesting for popular libs like immer to get involved.

**Comment by dckc** [comment]
@kriskowal in particular, you have made some progress getting some leading projects to adopt SES. Does it make sense to make that progress more visible to those looking at the proposal?

**Comment by kriskowal** [comment]
@dckc To make that progress more visible, I would need to make some progress. The next steps are to get the Compartments proposal ready for advancement, then produce a Lockdown/Harden proposal.

**Comment by dckc** [comment]
The progress I had in mind was stuff like getting SES into cosmjs. And some recent movement around babel?

_p.s. [What is the current status of SES, tinySES and Jessie? \- I have questions \- TC39](https://es.discourse.group/t/what-is-the-current-status-of-ses-tinyses-and-jessie/1306/5) is nice. [Propose to tc39 without override mistake · Issue \#105 · endojs/endo](https://github.com/endojs/endo/issues/105#issuecomment-1100245428) too._

**Comment by kriskowal** [comment]
That’s a good thought. The tracking issue for SES compatibility in COSMWASM is https://github.com/cosmos/cosmjs/issues/412.

**Comment by naugtur** [comment]
`__esModules` support as a flag to interpret `exports.default` content as the default when importing into ESM is not implemented.
This is also not supported by node.js according to the case `11.cjs default.even` in this table: https://github.com/endojs/endo-e2e-tests/blob/main/matrix/table.md  
Context and discussion: https://github.com/endojs/endo/issues/927

**Comment by erights** [comment]
See https://github.com/ReactiveX/rxjs/pull/6991

**Comment by erights** [comment]
See https://github.com/protobufjs/protobuf.js/pull/1742

Done!

**Comment by erights** [comment]
See https://github.com/nodejs/readable-stream/pull/474

Done!

**Comment by erights** [comment]
See https://github.com/zloirock/core-js/issues/1092

Addressed on the SES side at https://github.com/endojs/endo/pull/1221

So, done enough.

**Comment by michaelfig** [comment]
See nodejs/node#43496

See https://github.com/nodejs/node/pull/43907

Done!

**Comment by erights** [comment]
> Hey all, the error I get when trying to use `window.lockdown()` with react is
> 
> ```
> index.js:1 failed to delete intrinsics.Array.@wry/context:Slot of type function:  (TypeError#1)
> ```

@AidenRourke @benjamn @kumavis 
Given https://github.com/benjamn/wryware/blob/c490a8eef3f0976fec2a3ff8f939289f00e0a060/packages/context/src/slot.ts#L126 (thanks @benjamn )
is this issue with `Array.@wry` now fixed? Does `lockdown` now work in the presence of wryware?

**Comment by benjamn** [comment]
@erights @kumavis @rbuckton I'm responsible for the `@wry/` packages, and I'm happy to adjust anything you recommend.

To explain a bit, I originally chose the `Array["@wry/context:Slot"]` property of the `Array` constructor as a dirty but convenient place to stash a singleton instance of the `Slot` class, hoping that defining it as `enumerable: false` would prevent anyone from stumbling across it (in a bad way).

Regardless of how unsavory you consider that hack, I'm guessing I went too far in making the property also `writable: false` and `configurable: false`, since that seems to be what's preventing the deletion (from https://github.com/endojs/endo/issues/576#issuecomment-1178194624 above):
```
index.js:1 failed to delete intrinsics.Array.@wry/context:Slot of type function:  (TypeError#1)
```
I'm happy to make that change ASAP (loosening to `writable: true` and `configurable: true`), if we think merely enabling the deletion is an adequate answer here.

The `Slot` system still has a good chance of working if `Array["@wry/context:Slot"]` is lost for any reason (it just becomes more important to make sure there's only one copy of the `@wry/context` package in use), so I have no problem with you deleting this property in order to lock down the global environment, if that's a fair way to paraphrase your hopes here?

**Comment by ljharb** [comment]
Why not stash it on the global instead, behind a symbol when available, with spaces in the name so it doesn’t show up in repl autocomplete? (https://npmjs.com/global-cache does this, fwiw)

**Comment by erights** [comment]
> spaces in the name so it doesn’t show up in repl autocomplete

Cool! I did not know about that hack. 

Putting it on the initial global (with or without the space) is perfectly compatible with Hardened JS. It will not propagate by default to the global of constructed compartments, which is certainly ok by us. We already assume that the start compartment's global contains powerful things we do not know about. We do not remove or harden them by default.

**Comment by benjamn** [comment]
Alright, I think I might be able to use `globalThis` when available, instead of abusing `Array` as a namespace.

I've had many headaches over the years obtaining the global object in a reliable cross-browser/environment CSP-respecting way (people will yell at you if you resort to `Function("return this")()` without trying everything else first), but presumably the environments SES cares about all have `globalThis` at this point?

**Comment by erights** [comment]
> but presumably the environments SES cares about all have globalThis at this point?

Yes. (Attn @kriskowal )

**Comment by ljharb** [comment]
@benjamn fwiw global-cache uses `global`, which non-broken node module bundlers automatically should convert to `window` or  `globalThis` or `self` etc

**Comment by erights** [comment]
> Alright, I think I might be able to use `globalThis` when available, instead of abusing `Array` as a namespace.

Hi @benjamn , any progress on this? Anything we can do to help?

Filed https://github.com/benjamn/wryware/issues/347 so we can track this from your side too.

Done!

**Comment by erights** [comment]
The stack-utils packages trips the override mistake. Patches at https://github.com/tapjs/stack-utils/issues/70 and https://github.com/Agoric/agoric-sdk/pull/6451 . PR at https://github.com/tapjs/stack-utils/pull/71

Done!

**Comment by erights** [comment]
Immer

See https://github.com/immerjs/immer/pull/914 and https://github.com/endojs/endo/pull/1433

See messages starting at https://github.com/endojs/endo/issues/576#issuecomment-1090188459 above

**Comment by dckc** [comment]
When trying to import [google-sheets](https://www.npmjs.com/package/google-sheets), I hit `Cannot assign to read only property 'constructor'` somewhere in **axios**:

<details><summary>stack trace</summary>

```
TypeError#1: Cannot assign to read only property 'constructor' of object '[object Object]'
  at file:///home/connolly/projects/finquick/packages/fincaps/node_modules/axios/lib/utils.js:545:32
  at forEach (file:///home/connolly/projects/finquick/packages/fincaps/node_modules/axios/lib/utils.js:265:10)
  at Object.reduceDescriptors (file:///home/connolly/projects/finquick/packages/fincaps/node_modules/axios/lib/utils.js:542:3)
  at file:///home/connolly/projects/finquick/packages/fincaps/node_modules/axios/lib/core/AxiosHeaders.js:286:7
```
</details> 

The usual `Object.defineProperty` alternative seems to make the symptoms go away:

<details><summary>axios+1.5.0.patch</summary>

```patch
diff --git a/node_modules/axios/lib/utils.js b/node_modules/axios/lib/utils.js
index a386b77..ea8bbd6 100644
--- a/node_modules/axios/lib/utils.js
+++ b/node_modules/axios/lib/utils.js
@@ -542,7 +542,7 @@ const reduceDescriptors = (obj, reducer) => {
   forEach(descriptors, (descriptor, name) => {
     let ret;
     if ((ret = reducer(descriptor, name, obj)) !== false) {
-      reducedDescriptors[name] = ret || descriptor;
+      Object.defineProperty(reducedDescriptors, name, { value: ret || descriptor });
     }
   });
```
</details>

**Comment by erights** [comment]
Tracking https://github.com/rbuckton/reflect-metadata/issues/130 .

I should have added it to this list ages ago, but somehow forgot.

**Comment by mhofman** [comment]
I just hit some issues trying to get `agoric-cli` working without severe taming. 

<details>

<summary>Relevant stack traces:</summary>

```
(TypeError#1)
TypeError#1: Override property constructor
  at Error.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at createErrorType (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/follow-redirects/index.js:600:37)
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/follow-redirects/index.js:18:23)

(TypeError#6)
TypeError#6: Override property constructor
  at Object.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at newError (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/cosmjs-types/node_modules/protobufjs/src/util/minimal.js:283:74)
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/cosmjs-types/node_modules/protobufjs/src/util/minimal.js:313:22)

(TypeError#8)
TypeError#8: Override property constructor
  at Object.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at Object.createErrorClass (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/rxjs/dist/cjs/internal/util/createErrorClass.js:11:36)

(TypeError#15)
TypeError#15: Override property constructor
  at __.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at new __ (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/CreateFileError.js:17:42)
  at __extends (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/CreateFileError.js:18:84)
  at /tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/CreateFileError.js:23:5
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/CreateFileError.js:38:2)

(TypeError#16)
TypeError#16: Override property constructor
  at __.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at new __ (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/LaunchEditorError.js:17:42)
  at __extends (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/LaunchEditorError.js:18:84)
  at /tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/LaunchEditorError.js:23:5
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/LaunchEditorError.js:38:2)

(TypeError#17)
TypeError#17: Override property constructor
  at __.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at new __ (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/ReadFileError.js:17:42)
  at __extends (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/ReadFileError.js:18:84)
  at /tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/ReadFileError.js:23:5
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/ReadFileError.js:38:2)

(TypeError#18)
TypeError#18: Override property constructor
  at __.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at new __ (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/RemoveFileError.js:17:42)
  at __extends (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/RemoveFileError.js:18:84)
  at /tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/RemoveFileError.js:23:5
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/external-editor/main/errors/RemoveFileError.js:38:2)

(TypeError#19)
TypeError#19: Override property constructor
  at Object.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at _interopRequireWildcard (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/@babel/highlight/lib/index.js:12:710)
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/@babel/highlight/lib/index.js:10:14)

(TypeError#20)
TypeError#20: Override property constructor
  at Object.setter (file:///tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/ses/src/enable-property-overrides.js:116:27)
  at _interopRequireWildcard (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/@babel/code-frame/lib/index.js:11:710)
  at Object.<anonymous> (/tmp/registry-home.UVpYi/.npm/_npx/417013ce1422aeac/node_modules/@babel/code-frame/lib/index.js:9:14)
```

</details>

Issues already mitigated by `moderate` taming:
- https://github.com/follow-redirects/follow-redirects/issues/231: An issue in [follow-redirects](https://github.com/follow-redirects/follow-redirects/blob/192dbe7ce671ecad813c074bffe3b3f5d3680fee/index.js#L600-L601)
- https://github.com/confio/cosmjs-types/issues/78: [cosmjs-types](https://github.com/confio/cosmjs-types/blob/073883429ae6b2c311bbcf2970a7bc7219f3e2f1/package.json#L43C6-L43C16) that need to update to version ^7 of protobufjs to pick up https://github.com/protobufjs/protobuf.js/pull/1742
- known [rxjs](https://github.com/ReactiveX/rxjs/pull/6991) issue that needs to be re-opened by contributing tests (and verify no new cases have been introduced)
- Known unresolved issue https://github.com/mrkmg/node-external-editor/issues/25

New issue requiring taming of `%ObjectPrototype%.constructor`:
- https://github.com/babel/babel/issues/15995: An override mistake in [`interopRequireWildcard` of `@babel/helpers`](https://github.com/babel/babel/blob/4a0af42f571905dcca688be5c5ca8e032554fdc9/packages/babel-helpers/src/helpers.ts#L407-L410), which is triggered by the package `chalk` having a weird [own enumerable `constructor` property](https://github.com/chalk/chalk/blob/v2.4.2/index.js#L42C38-L42C38), that is [used with named imports](https://github.com/babel/babel/blob/v7.22.20/packages/babel-highlight/src/index.ts#L11) in the latest versions of `@babel/highlight` and `@babel/code-frame`

**Comment by erights** [comment]
See https://github.com/endojs/endo/issues/1846

**Comment by kriskowal** [comment]
chalk@4 all versions https://github.com/chalk/chalk/issues/619

**Comment by erights** [comment]
See https://github.com/rbuckton/reflect-metadata/pull/131 and https://github.com/rbuckton/reflect-metadata/pull/144

**Comment by erights** [comment]
From https://github.com/endojs/endo/pull/2030 👍 

> The problem fixed by this PR is interesting. The override mistake is in mobx generated code, where the corresponding mobx source code is a TypeScript class definition. The generated code is not a JS class definition, but rather, seems to target pre-class JS. The surprising mistake is that the generated code is initializing the methods on the "class" .prototype using assignment rather than `defineProperty` in blatant violation of the JS class semantics. Presumably, this is also a blatant violation of the current TS class semantics, but I don't know. Is some ancient TS compiler at play here? Or some non-standard TS or TS-like compiler? I don't know.

**Comment by kriskowal** [comment]
Please post updates to https://github.com/endojs/endo/issues/2037
